WalletWasabi/Bases/PeriodicRunner.cs

```diff
@@ -1,132 +1,97 @@
+			_tcs?.TrySetResult(true);
 		}
 
+		/// <summary>
+		/// Abstract method that is called every <see cref="Period"/> or sooner when <see cref="TriggerRound"/> is called.
+		/// </summary>
 		protected abstract Task ActionAsync(CancellationToken cancel);
 
 		/// <inheritdoc />
 		protected override async Task ExecuteAsync(CancellationToken stoppingToken)
 		{
 			while (!stoppingToken.IsCancellationRequested)
 			{
+				_tcs = new TaskCompletionSource<bool>();
+
 				try
 				{
+					// Do user action.
 					await ActionAsync(stoppingToken).ConfigureAwait(false);
-					LogAndResetLastExceptionIfNotNull();
+
+					// Log previous exception if any.
+					LogLastException(ExceptionTracker.LastException);
 				}
 				catch (Exception ex) when (ex is OperationCanceledException || ex is TaskCanceledException || ex is TimeoutException)
 				{
 					Logger.LogTrace(ex);
 				}
 				catch (Exception ex)
 				{
-					// Only log one type of exception once.
-					if (LastException is null // If the exception never came.
-						|| ex.GetType() != LastException.GetType() // Or the exception have different type from previous exception.
-						|| ex.Message != LastException.Message) // Or the exception have different message from previous exception.
-					{
-						// Then log and reset the last exception if another one came before.
-						LogAndResetLastExceptionIfNotNull();
-						// Set new exception and log it.
-						LastException = ex;
-						LastExceptionFirstAppeared = DateTimeOffset.UtcNow;
-						LastExceptionCount = 1;
-						Logger.LogError(ex);
-					}
-					else
-					{
-						// Increment the exception counter.
-						LastExceptionCount++;
-					}
+					// Exception encountered. Process it and log it, if it is first in the row.
+					LogLastException(ExceptionTracker.Process(ex));
+
+					Logger.LogError(ExceptionTracker.LastException!.Exception);
+				}
+
+				// Wait for the next round.
+				try
+				{
+					await Task.WhenAny(_tcs.Task, Task.Delay(Period, stoppingToken)).ConfigureAwait(false);
 				}
-				finally
+				catch (TaskCanceledException ex)
 				{
-					try
-					{
-						CancellationToken? triggeringToken = null;
-						lock (TriggerLock)
-						{
-							triggeringToken = TriggeringCts?.Token;
-						}
-						if (triggeringToken is { })
-						{
-							using var linked = CancellationTokenSource.CreateLinkedTokenSource(stoppingToken, triggeringToken.Value);
-							await Task.Delay(Period, linked.Token).ConfigureAwait(false);
-						}
-					}
-					catch (TaskCanceledException ex)
-					{
-						Logger.LogTrace(ex);
-					}
-					finally
-					{
-						lock (TriggerLock)
-						{
-							if (TriggeringCts?.IsCancellationRequested is true)
-							{
-								TriggeringCts?.Dispose();
-								TriggeringCts = new CancellationTokenSource();
-							}
-						}
-					}
+					Logger.LogTrace(ex);
 				}
 			}
 		}
 
-		private void LogAndResetLastExceptionIfNotNull()
+		private void LogLastException(ExceptionInfo? info)
```

Comment:
All the code would look better if we don't pass nulls.

Reply:
I'm not sure how to achieve that - can you elaborate how to do that? 

I use `ExceptionInfo?` precisely because an exception may have occurred or not. `LastExceptionTracker` was designed so that it does not contain any call of `Logger` so that it's potentially re-usable.